### ISSUE TITLE
Broadcasting breaks scripts and tests

### ISSUE THREAD

**Issue by PatrickAlphaC** [matter-labs/foundry-zksync#389]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.0.2 (dc4b81f 2024-05-21T15:26:16.686968000Z)

### What command(s) is the bug in?

forge test

### Operating System

None

### Describe the bug

If you run `vm.startBroadcast();` in a script, forge tests will break.

# Example

1. Setup a new project
```
forge init
```
2. Change the following files:

`Counter.s.sol`

```javascript
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

import {Script, console} from "forge-std/Script.sol";
import {Counter} from "src/Counter.sol";

contract CounterScript is Script {
    function setUp() public {}

    function run() public returns (Counter) {
        vm.startBroadcast();
        Counter counter = new Counter();
        vm.stopBroadcast();
        return counter;
    }
}
```

`Counter.t.sol`

```javascript
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

import {Test, console} from "forge-std/Test.sol";
import {Counter} from "../src/Counter.sol";
import {CounterScript} from "script/Counter.s.sol";

contract CounterTest is Test {
    Counter public counter;
    CounterScript public script;

    function setUp() public {
        script = new CounterScript();
        counter = script.run();
    }

    function test_Increment() public {
        counter.increment();
        assertEq(counter.number(), 1);
    }
}
```

3. Run a test

```
forge test --mt test_Increment --zksync -vvvv
```

You'll get an `EvmError: Revert` during the setup. You'll notice if you do this with vanilla foundry after running `foundryup` it does not fail. 

```
foundryup
forge test --mt test_Increment -vvvv
```

**Comment by Karrq** [comment]
Hey @PatrickAlphaC, the team has discussed the issue at hand. 
Your issue fundamentally clashes against how we approach handling EraVM and EVM together during test execution.

In particular, your script's `run` method would be executed inside the EraVM, where cheatcodes are not enabled, which leads to the failure you described, as the `vm` address isn't handled, nor any of the cheatcodes.

I understand your intent in avoiding code duplication, but you'll need to restructure it as adding support for this functionality would mean almost redoing our EraVM integration in its entirety.

TL;DR
Running cheatcodes inside EraVM (any contract call inside the test) is not supported.

PS: the `broadcast`-family of cheatcodes is rather useless in tests as they don't accomplish anything in that context.

**Comment by PatrickAlphaC** [comment]
Why would this change "almost redoing our EraVM integration in its entirety"?

Why not just have a flag that says "if local tests, ignore that cheatcode"? 

It seems developers (or perhaps, me at least) are running into more and more issues where the devex on zkSync is quite different than other EVM chains. This should be avoided if we would like people to deploy to zkSync. Otherwise, you're asking people to have separate environments and test suites for zkSync, which seems to defeat the purpose of having solidity compatibility.
